### DISCUSSION TITLE
circular import

### DISCUSSION THREAD

**Discussion by meng-jpg** [eclipse-theia/theia#10649] [Q&A] [answered]
I'm not sure whether this will cause some problems?

This is a `circular import`:
The class `ShellLayoutRestorer` is imported by `frontend-application.ts` and used as dependency of the class `FrontendApplication`.
At the same time, `FrontendApplication` is imported and used in class `ShellLayoutRestorer`.

A second one：
`shell-layout-register.ts` import `CommonCommands` from `common-frontend-contribution.ts`.
`common-frontend-contribution.ts` import some classes from `frontend-application.ts`.
`frontend-application.ts` import `ShellLayoutRestorer` from `shell-layout-register.ts`.

<img width="744" alt="截屏2022-01-19 下午4 08 25" src="https://user-images.githubusercontent.com/74631407/150091268-42355666-0a45-4e4f-bdf4-052fff51c362.png">
<img width="868" alt="截屏2022-01-19 下午4 16 33" src="https://user-images.githubusercontent.com/74631407/150091291-3b07be60-88c5-43a9-aafd-f02cc184d123.png">
<img width="939" alt="截屏2022-01-19 下午4 16 03" src="https://user-images.githubusercontent.com/74631407/150091278-2001529e-4312-44e7-921b-8a9451db8101.png">

code version: theia 1.21.0

**Comment by vince-fugnitto** [comment, marked as answer]
> I'm not sure whether this will cause some problems?

@meng-jpg thank you for the discussion, in general we try and avoid circular dependencies (although sometimes it is difficult and may lead to breaking changes). We fixed many in the past that were problematic (prevented the app from starting properly), but some are not necessarily issues.

In any case, if you want to fix some of the issues please feel free and go ahead, pull-request are always welcome :+1: 
One tool I like to use to help identify them is [madge](https://www.npmjs.com/package/madge).
